File: src/main/java/com/nuutrai/reactor/util/FileUtils.java
package com.nuutrai.reactor.util;

import java.io.File;
import java.io.IOException;
import java.nio.file.FileVisitResult;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.SimpleFileVisitor;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Stream;

import static com.nuutrai.reactor.Reactor.logger;
import static java.nio.file.StandardCopyOption.COPY_ATTRIBUTES;

public class FileUtils {

    public static boolean deleteFolder(File file) {
        try (Stream<Path> files = Files.walk(file.toPath())) {
            files.sorted(Comparator.reverseOrder()).map(Path::toFile).forEach(File::delete);
            return true;
        } catch (IOException e) {
            logger.warning(e.getMessage());
            return false;
        }
    }

    public static boolean copyFolder(File source, File target, List<String> excludeFiles) {
        Path sourceDir = source.toPath();
        Path targetDir = target.toPath();

        try {
            Files.walkFileTree(sourceDir, new CopyDirFileVisitor(sourceDir, targetDir, excludeFiles));
            return true;
        } catch (IOException e) {
            logger.warning("Unable to copy directory " + e);
            return false;
        }
    }

    private static class CopyDirFileVisitor extends SimpleFileVisitor<Path> {

        private final Path sourceDir;
        private final Path targetDir;
        private final List<String> excludeFiles;

        private CopyDirFileVisitor(Path sourceDir, Path targetDir, List<String> excludeFiles) {
            this.sourceDir = sourceDir;
            this.targetDir = targetDir;
            this.excludeFiles = excludeFiles;
        }

        @Override
        public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
            Path newDir = targetDir.resolve(sourceDir.relativize(dir));
            if (!Files.isDirectory(newDir)) {
                Files.createDirectory(newDir);
            }
            return FileVisitResult.CONTINUE;
        }

        @Override
        public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
            // Pass files that are set to ignore
            if (excludeFiles != null && excludeFiles.contains(file.getFileName().toString()))
                return FileVisitResult.CONTINUE;
            // Copy the files
            Path targetFile = targetDir.resolve(sourceDir.relativize(file));
            Files.copy(file, targetFile, COPY_ATTRIBUTES);
            return FileVisitResult.CONTINUE;
        }
    }

}

File: src/main/kotlin/com/nuutrai/reactor/util/FileUtils.kt
package com.nuutrai.reactor.util

import com.nuutrai.reactor.Reactor
import java.io.File
import java.io.IOException
import java.nio.file.*
import java.nio.file.attribute.BasicFileAttributes

object FileUtils {
	fun deleteFolder(file: File): Boolean {
		try {
			Files.walk(file.toPath()).use { files ->
				files.sorted(Comparator.reverseOrder<Path?>()).map<File?> { obj: Path? -> obj!!.toFile() }
					.forEach { obj: File? -> obj!!.delete() }
				return true
			}
		} catch (e: IOException) {
			Reactor.Companion.logger!!.warning(e.message)
			return false
		}
	}

	fun copyFolder(source: File, target: File, excludeFiles: MutableList<String?>?): Boolean {
		val sourceDir = source.toPath()
		val targetDir = target.toPath()

		try {
			Files.walkFileTree(sourceDir, CopyDirFileVisitor(sourceDir, targetDir, excludeFiles))
			return true
		} catch (e: IOException) {
			Reactor.Companion.logger!!.warning("Unable to copy directory " + e)
			return false
		}
	}

	private class CopyDirFileVisitor(
		private val sourceDir: Path,
		private val targetDir: Path,
		private val excludeFiles: MutableList<String?>?
	) : SimpleFileVisitor<Path?>() {
		@Throws(IOException::class)
		override fun preVisitDirectory(dir: Path, attrs: BasicFileAttributes?): FileVisitResult {
			val newDir = targetDir.resolve(sourceDir.relativize(dir))
			if (!Files.isDirectory(newDir)) {
				Files.createDirectory(newDir)
			}
			return FileVisitResult.CONTINUE
		}

		@Throws(IOException::class)
		override fun visitFile(file: Path, attrs: BasicFileAttributes?): FileVisitResult {
			// Pass files that are set to ignore
			if (excludeFiles != null && excludeFiles.contains(
					file.getFileName().toString()
				)
			) return FileVisitResult.CONTINUE
			// Copy the files
			val targetFile = targetDir.resolve(sourceDir.relativize(file))
			Files.copy(file, targetFile, StandardCopyOption.COPY_ATTRIBUTES)
			return FileVisitResult.CONTINUE
		}
	}
}
